### ISSUE TITLE
Cannot deploy V2 Cloud Functions with more than 8 GiB of memory

### ISSUE THREAD

**Issue by 0xjei** [firebase/firebase-tools#4803]
### [REQUIRED] Environment info

**firebase-tools**: *^11.4.1*

**Platform**: *Ubuntu*

### [REQUIRED] Test case

You can reproduce the error by defining a V2 Cloud Function as below:

```javascript
import { onCall } from 'firebase-functions/v2/https'

export const dummycf = onCall({
  timeoutSeconds: 3600,
  memory: "16GiB",
  cpu: 2
}, (request) => {
  console.log(`This V2 CF w/ 16GiB of memory and 2 vCPUs should work fine!`)
})
```

### [REQUIRED] Steps to reproduce

You just need to deploy the Cloud Function by running:

```bash 
firebase deploy --only functions:test
```

### [REQUIRED] Expected behavior

The Cloud Function must be correctly deployed with 4 vCPUs and 16 GiB of memory. The V2 public preview should support the memory and cpu values as reported in the [documentation](https://firebase.google.com/docs/functions/beta/manage-functions).

### [REQUIRED] Actual behavior

Running the command above results in the following error:

```bash
> firebase deploy --only functions:dummycf

=== Deploying to 'mpc-phase2-suite-test'...

i  deploying functions
Running command: npm --prefix "$RESOURCE_DIR" run build

> mpc-phase2-suite-backend@0.0.1 build
> tsc

✔  functions: Finished running predeploy script.
i  functions: ensuring required API cloudfunctions.googleapis.com is enabled...
i  functions: ensuring required API cloudbuild.googleapis.com is enabled...
i  artifactregistry: ensuring required API artifactregistry.googleapis.com is enabled...
✔  artifactregistry: required API artifactregistry.googleapis.com is enabled
✔  functions: required API cloudbuild.googleapis.com is enabled
✔  functions: required API cloudfunctions.googleapis.com is enabled
i  functions: preparing codebase default for deployment
i  functions: Loaded environment variables from .env.

Error: available memory must be a supported value, if present
```

**Comment by taeold** [comment]
Hey @0xjei - thanks for reporting and apologies for the issue you are seeing.

I tried giving 16GiB function a spin and it gave my setup no problem:

```
exports.sixteen = functionsv2.https.onCall({ memory: "16GiB", cpu: 2 }, () => { return { foo: "bar"}; });
```

![image](https://user-images.githubusercontent.com/14897098/182496406-e7d19800-2a29-44bf-8c8e-949d66ecc518.png)

Can you check to see if you are using the latest version of the firebase-functions SDK? In my setup, I was using 3.22.0.

**Comment by taeold** [comment]
@0xjei Yes it should be working in both cases. Thank you for your detailed feedback - I'll go back and see what I can uncover.

**Comment by taeold** [comment]
I played around with several settings, and at least locally on my machine:

```
import { onCall } from "firebase-functions/v2/https";
export const shake = onCall({ memory: "16GiB"}, () => "hello");
```

```
import * as fv2 from "firebase-functions/v2";
export const wild = fv2.https.onCall({ memory: "16GiB" }, () => "hello");
```

Both successfully deployed a 16 GiB function:

![image](https://user-images.githubusercontent.com/14897098/182666748-c345203c-e785-49e6-9b8a-ebe25f43ccec.png)

That said, I think there are several inconsistencies that can be introduced in the version of the Firebase CLI and Functions SDK isn't compatible. For example, order versions of the Functions SDK or CLI won't support 16GiB memory option and might lead to inconsistent deployment behavior.

There are rough edges here - I think we will have to, at some point, make a breaking change in the CLI to emit errors like "Please use Functions SDK version > X.X.X" instead of leaving things in consistent state as you've helped us discover. Please stay tuned.

Until then - do you mind double checking your usage with `import { onCall } from "firebase-functions/v2/https";` style and letting me know how it goes. It would be great if you can share a little more about your setup (e.g. if you are using TypeScript, sharing some parts of `tsconfig.json` would be super helpful for debugging since I can't reproduce the error yet).

**Comment by 0xjei** [comment]
Hi @taeold! I appreciate the effort and clarity of your response. Anyway, I may have found the problem here and I think this is closely related to this PR #4823. 

As you know, I have been trying different combinations of memory + cpu for CF V2 these days without success. When I found this PR, I noticed that  when I was importing the `onCall()` method I was configuring the function with just 16 GiB as memory and no CPU option set, leading to the invalid CPU setting error that you've found in the PR. 

So, just adding a `cpu: 4` option the deploy with both import styles works fine! :tada: 

Anyway, something strange happens when I deploy the functions for the __first__ time: everything seems correct because I can see the functions in the dashboard without no errors and when I call them everything works fine and I can see logs and results. However, the deploy script got me back this error:

```bash
npx firebase deploy --only functions:helloworld

=== Deploying to 'mpc-phase2-suite-test'...

i  deploying functions
Running command: yarn --prefix "$RESOURCE_DIR" build
yarn run v1.22.18
$ tsc
Done in 2.12s.
✔  functions: Finished running predeploy script.
i  functions: ensuring required API cloudfunctions.googleapis.com is enabled...
i  functions: ensuring required API cloudbuild.googleapis.com is enabled...
i  artifactregistry: ensuring required API artifactregistry.googleapis.com is enabled...
✔  functions: required API cloudbuild.googleapis.com is enabled
✔  artifactregistry: required API artifactregistry.googleapis.com is enabled
✔  functions: required API cloudfunctions.googleapis.com is enabled
i  functions: preparing codebase default for deployment
i  functions: Loaded environment variables from .env.
i  functions: preparing . directory for uploading...
i  functions: packaged /code/mpc-phase2-suite/firebase (227.87 KB) for uploading
i  functions: packaged /code/mpc-phase2-suite/firebase (228.1 KB) for uploading
i  functions: ensuring required API cloudscheduler.googleapis.com is enabled...
✔  functions: required API cloudscheduler.googleapis.com is enabled
i  functions: ensuring required API run.googleapis.com is enabled...
i  functions: ensuring required API eventarc.googleapis.com is enabled...
i  functions: ensuring required API pubsub.googleapis.com is enabled...
i  functions: ensuring required API storage.googleapis.com is enabled...
✔  functions: required API pubsub.googleapis.com is enabled
✔  functions: required API eventarc.googleapis.com is enabled
✔  functions: required API storage.googleapis.com is enabled
✔  functions: required API run.googleapis.com is enabled
i  functions: generating the service identity for pubsub.googleapis.com...
i  functions: generating the service identity for eventarc.googleapis.com...
✔  functions: . folder uploaded successfully
i  functions: creating Node.js 16 function helloworld(us-central1)...
Failed to update Run service projects/mpc-phase2-suite-test/locations/us-central1/services/helloworld

Functions deploy had errors with the following functions:
        helloworld(us-central1)
i  functions: cleaning up build files...

Error: There was an error deploying functions
```

And you can see more clearly the error from the debug logs:

```log
// Everything seems okay before this line...
[error] Failed to update Run service projects/mpc-phase2-suite-test/locations/us-central1/services/helloworld
[debug] [2022-08-04T10:19:48.823Z] Total Function Deployment time: 198044
[debug] [2022-08-04T10:19:48.823Z] 1 Functions Deployed
[debug] [2022-08-04T10:19:48.823Z] 1 Functions Errored
[debug] [2022-08-04T10:19:48.823Z] 0 Function Deployments Aborted
[debug] [2022-08-04T10:19:48.823Z] Average Function Deployment time: 198043
[info] 
[info] Functions deploy had errors with the following functions:
	helloworld(us-central1)
[info] i  functions: cleaning up build files... 
[debug] [2022-08-04T10:19:48.835Z] >>> [apiv2][query] DELETE https://artifactregistry.googleapis.com/v1beta2/projects/mpc-phase2-suite-test/locations/us-central1/repositories/gcf-artifacts/packages/helloworld [none]
[debug] [2022-08-04T10:19:49.331Z] Could not find image for function projects/mpc-phase2-suite-test/locations/us-central1/functions/helloworld
[debug] [2022-08-04T10:19:49.752Z] <<< [apiv2][status] DELETE https://artifactregistry.googleapis.com/v1beta2/projects/mpc-phase2-suite-test/locations/us-central1/repositories/gcf-artifacts/packages/helloworld 200
[debug] [2022-08-04T10:19:50.457Z] Functions deploy failed.
[debug] [2022-08-04T10:19:50.457Z] {
  "endpoint": {
    "id": "helloworld",
    "project": "mpc-phase2-suite-test",
    "region": "us-central1",
    "entryPoint": "helloworld",
    "platform": "gcfv2",
    "runtime": "nodejs16",
    "callableTrigger": {},
    "labels": {
      "deployment-tool": "cli-firebase"
    },
    "availableMemoryMb": 32768,
    "environmentVariables": {
      "FIRESTORE_DATABASE_URL": "....firebaseio.com", // not publicly available
      "PROJECT_ID": "...", not publicly available
      "STORAGE_BUCKET": "....appspot.com", // not publicly available
      "SCHEDULER_START_CEREMONY": "20",
      "SCHEDULER_STOP_CEREMONY": "20",
      "FIRST_ZKEY_INDEX": "00000",
      "TIMEOUT_TOLERANCE_RATE": "20",
      "RETRY_WAITING_TIME_IN_DAYS": "1",
      "CUSTOM_CLAIMS_COORDINATOR_EMAIL_ADDRESS_OR_DOMAIN": "...@gmail.com", // not publicly available
      "FIREBASE_CONFIG": "{\"projectId\":\"...\",\"storageBucket\":\"....appspot.com\",\"locationId\":\"...\"}", not publicly available
      "GCLOUD_PROJECT": "..." not publicly available
    },
    "codebase": "default",
    "cpu": 4,
    "uri": "https://helloworld-43d5rlynya-uc.a.run.app",
    "concurrency": 80
  },
  "op": "set concurrency",
  "original": {
    "name": "FirebaseError",
    "children": [],
    "exit": 1,
    "message": "Failed to update Run service projects/mpc-phase2-suite-test/locations/us-central1/services/helloworld",
    "original": {
      "name": "FirebaseError",
      "children": [],
      "context": {
        "body": {
          "error": {
            "code": 400,
            "message": "spec.template.spec.containers[0].resources.limits.memory: Invalid value specified for memory. For 4.0 CPU, memory must be between 2Gi and 16Gi inclusive.\nFor more troubleshooting guidance, see https://cloud.google.com/run/docs/configuring/memory-limits",
            "status": "INVALID_ARGUMENT",
            "details": [
              {
                "@type": "type.googleapis.com/google.rpc.BadRequest",
                "fieldViolations": [
                  {
                    "field": "spec.template.spec.containers[0].resources.limits.memory",
                    "description": "Invalid value specified for memory. For 4.0 CPU, memory must be between 2Gi and 16Gi inclusive.\nFor more troubleshooting guidance, see https://cloud.google.com/run/docs/configuring/memory-limits"
                  }
                ]
              }
            ]
          }
        },
        "response": {
          "statusCode": 400
        }
      },
      "exit": 1,
      "message": "HTTP Error: 400, spec.template.spec.containers[0].resources.limits.memory: Invalid value specified for memory. For 4.0 CPU, memory must be between 2Gi and 16Gi inclusive.\nFor more troubleshooting guidance, see https://cloud.google.com/run/docs/configuring/memory-limits",
      "status": 400
    },
    "status": 500
  }
}
[debug] [2022-08-04T10:19:50.458Z] Error: Failed to set concurrency function helloworld in region us-central1
    at /code/mpc-phase2-suite/node_modules/firebase-tools/lib/deploy/functions/release/fabricator.js:41:11
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Fabricator.setRunTraits (/code/mpc-phase2-suite/node_modules/firebase-tools/lib/deploy/functions/release/fabricator.js:388:9)
    at async Fabricator.createV2Function (/code/mpc-phase2-suite/node_modules/firebase-tools/lib/deploy/functions/release/fabricator.js:283:13)
    at async Fabricator.createEndpoint (/code/mpc-phase2-suite/node_modules/firebase-tools/lib/deploy/functions/release/fabricator.js:121:13)
    at async handle (/code/mpc-phase2-suite/node_modules/firebase-tools/lib/deploy/functions/release/fabricator.js:77:17)
[error] 
[error] Error: There was an error deploying functions

```

On the other hand, when I "update" them (i.e., function is deployed and I upgrade the memory or CPU settings), I get no error as you can noticed from the "update" log:

```bash
npx firebase deploy --only functions:helloworld

=== Deploying to 'mpc-phase2-suite-test'...

i  deploying functions
Running command: yarn --prefix "$RESOURCE_DIR" build
yarn run v1.22.18
$ tsc
Done in 2.14s.
✔  functions: Finished running predeploy script.
i  functions: ensuring required API cloudfunctions.googleapis.com is enabled...
i  functions: ensuring required API cloudbuild.googleapis.com is enabled...
i  artifactregistry: ensuring required API artifactregistry.googleapis.com is enabled...
✔  functions: required API cloudfunctions.googleapis.com is enabled
✔  artifactregistry: required API artifactregistry.googleapis.com is enabled
✔  functions: required API cloudbuild.googleapis.com is enabled
i  functions: preparing codebase default for deployment
i  functions: Loaded environment variables from .env.
i  functions: preparing . directory for uploading...
i  functions: packaged /code/mpc-phase2-suite/firebase (227.87 KB) for uploading
i  functions: packaged /code/mpc-phase2-suite/firebase (228.1 KB) for uploading
i  functions: ensuring required API cloudscheduler.googleapis.com is enabled...
✔  functions: required API cloudscheduler.googleapis.com is enabled
i  functions: ensuring required API run.googleapis.com is enabled...
i  functions: ensuring required API eventarc.googleapis.com is enabled...
i  functions: ensuring required API pubsub.googleapis.com is enabled...
i  functions: ensuring required API storage.googleapis.com is enabled...
✔  functions: required API eventarc.googleapis.com is enabled
✔  functions: required API storage.googleapis.com is enabled
✔  functions: required API pubsub.googleapis.com is enabled
✔  functions: required API run.googleapis.com is enabled
i  functions: generating the service identity for pubsub.googleapis.com...
i  functions: generating the service identity for eventarc.googleapis.com...
✔  functions: . folder uploaded successfully
i  functions: updating Node.js 16 function helloworld(us-central1)...
✔  functions[helloworld(us-central1)] Successful update operation.
i  functions: cleaning up build files...

✔  Deploy complete!

Project Console: https://console.firebase.google.com/project/mpc-phase2-suite-test/overview
```

Also, I agree w/ you in making the user aware of the SDK upgrades "Please use Functions SDK version > X.X.X". I can add something too here: updating the `firebase-admin` package to the latest version (v11.0.1 rn) helped me to avoid inconsistencies when passing memory and CPU configs (because I'm using the library to initialize the app). 

I hope it can help you to close the loop and thanks again for the support and patience :1st_place_medal:

**Comment by taeold** [comment]
@0xjei Thanks again for detailed bug report and your patience being on the bleeding edge. Will get back at the create issue soon,

**Comment by taeold** [comment]
@0xjei Sorry I didn't get back earlier.

I reproduced a similar error to yours in my local setup:

```
import { onRequest } from "firebase-functions/v2/https";
export const v = onRequest({ memory: "16GiB", cpu: 2  }, (req, res) => { res.sendStatus(200); });
```

16GiB + 2 CPU. I get following error:

```
Invalid value specified for memory. For 2.0 CPU, memory must be between 128Mi and 8Gi inclusive.

For more troubleshooting guidance, see https://cloud.google.com/run/docs/configuring/memory-limits
```

So it turns out 2 CPU + 16 GiB is not an allowed configuration for Cloud Run. This is a surprised - I don't see anything about **maximum* memory per CPU shape.

Looking again at your error message:

 ```
Invalid value specified for memory. For 4.0 CPU, memory must be between 2Gi and 16Gi inclusive.
```

Looks like your function is configured w/ 32 GiB which isn't allowed to have 4.0 CPU - it must have 8.0 CPU or more.

I think we still have couple of bugs on incorrect mapping of memory -> CPU that I'm going to try to fix here. Thanks for helping us find the bug.

(PS: I'm not sure why the subsequent update is working for you 🤔. I'm going to focus on fixing the bug for now)

**Comment by OmerRosenbaum** [comment]
Hi, I'm also having a similar issue - I have `firebase-functions` version `3.22.0`, and I am running this:
```
export const codeanalysisv2 = functionsV2.https.onCall(
  codeAnalysisCloudFunctionRuntimeOpts,
...
```
Where:
```
const codeAnalysisCloudFunctionRuntimeOpts: HttpsOptions = {
  timeoutSeconds: 1800,
  memory: '16GiB',
  cpu: 4,
}

And I am getting:
`Error: available memory must be a supported value, if present`

How should I debug this?
Thank you!
```

**Comment by 0xjei** [comment]
Hey, @taeold! Thank you for your explanations and responsiveness in solving the issue. Glad to have helped you guys finding the bug. 

BTW, after your fix and upgrading to `11.7.0` version everything seems working fine when deploying combinations of 16GiB/4CPUs and 32GiB/8CPUs :tada:

**Comment by taeold** [comment]
Great to hear! Thanks for being proactive and helping up squash many bugs! 🐛
